working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+add support for _coroutines_ (also known as _fibers_, or _lightweight threads_)
+as well.
+
+```dart
+abstract interface class Coroutine {
+  /// Return currently running coroutine if any.
+  static Coroutine? get current;
+
+  /// Create a suspended coroutine which will execute the given
+  /// [body] when resumed.
+  static Coroutine create(void Function() body);
+
+  /// Suspends the given currently running coroutine.
+  ///
+  /// This makes `resume` return with
+  /// Expects resumer to pass back a value of type [R].
+  static void suspend();
+
+ 	/// Resumes previously suspended coroutine.
+  ///
+  /// If there is a coroutine currently running the suspends it
+  /// first.
+  void resume();
+
+  /// Resumes previously suspended coroutine with exception.
+  void resumeWithException(Object error, [StackTrace? st]);
+}
+```
+
+Coroutines is a very powerful abstraction which allows to write straight-line
+code which depends on asynchronous values.
+
+```dart
+Future<String> request(String uri);
+
+extension FutureSuspend<T> on Future<T> {
+  T get value {
+    final cor = Coroutine.current ?? throw 'Not on a coroutine';
+    late final T value;
+    this.then((v) {
+      value = v;
+      cor.resume();
+    }, onError: cor.resumeWithException);
+    cor.suspend();
+    return value;
+  }
+}
+
+List<String> requestAll(List<String> uris) =>
+  Future.wait(uris.map(request)).value;
+
+SomeResult processUris(List<String> uris) {
+	final data = requestAll(uris);
+  // some processing of [data]
+  // ...
+}
+
+void main() {
+  final uris = [...];
+  Coroutine.create(() {
+    final result = processUris(uris);
+    print(result);
+  }).resume();
+}
+```
+
+#### Blocking operations
+
+It might be useful to augment existing types like `Future`, `Stream` and
+`ReceivePort` with blocking APIs which would only be usable in shared isolate
+and under condition that it is not going to block the executor's event loop.
+
+### `dart:ffi` updates
+
+`dart:ffi` should expose atomic reads and writes for native memory.
+
+```dart
+extension Int32PointerAtomics on Pointer<Int32> {
+  void atomicStore(int value);
+  int atomicLoad();
+  (bool, int) compareAndSwap(int expected, int desired);
+  int fetchAdd(int v);
+  int fetchSub(int v);
+  int fetchAnd(int v);
+  int fetchOr(int v);
+  int fetchXor(int v);
+}
+
+extension IntPtrPointerAtomics on Pointer<IntPtr> {
+  void atomicStore(int value);
+  int atomicLoad();
+  (bool, int) compareAndSwap(int expected, int desired);
+  int fetchAdd(int v);
+  int fetchSub(int v);
+  int fetchAnd(int v);
+  int fetchOr(int v);
+  int fetchXor(int v);
+}
+
+extension PointerPointerAtomics<T> on Pointer<Pointer<T>> {
+  void atomicStore(Pointer<T> value);
+  Pointer<T> atomicLoad();
+  (bool, Pointer<T>) compareAndSwap(Pointer<T> expected, Pointer<T> desired);
+}
+```
+
+For convenience reasons we might also consider making the following work:
+
+```dart
+final class MyStruct extends Struct {
```

Comment:
C++ doesn't have a stable ABI, but it would be kinda of corresponding to

```c++
struct MyStruct {
    std::atomic<int32_t> value;
}
```

I'd need to check if the memory layout of std::atomic is identical to not using atomic and it only influences code gen, I'd expect so. If so, we might even explore generating these in FFIgen.

Reply:
I don't mean to imply that this will be a direct equivalent of `std::atomic<int32_t>` or `atomic_int` / `_Atomic(int)` from C11 - because as you say these don't actually have an defined stable ABI.

This is just a convenient way to define atomic fields in structs. 

Added a note about this.